File: internal/pkg/dto/request/sales_lead.go
package request

import (
	"agent-desk/internal/pkg/enums"
	"time"
)

type LeadData struct {
	Title       string          `json:"title"`
	Product     string          `json:"product"`
	Quantity    string          `json:"quantity"`
	Destination string          `json:"destination"`
	Budget      string          `json:"budget"`
	Timeline    string          `json:"timeline"`
	ContactName string          `json:"contactName"`
	Company     string          `json:"company"`
	Email       string          `json:"email"`
	Phone       string          `json:"phone"`
	Needs       string          `json:"needs"`
	AutoTags    []enums.LeadTag `json:"autoTags"`
}
type UpdateSalesLead struct {
	Revision       int64            `json:"revision"`
	Data           LeadData         `json:"data"`
	Status         enums.LeadStatus `json:"status"`
	OwnerID        int64            `json:"ownerId"`
	FollowUpAt     *time.Time       `json:"followUpAt"`
	Note           string           `json:"note"`
	CustomTags     []string         `json:"customTags"`
	AcceptProposal bool             `json:"acceptProposal"`
}

type FollowUpSalesLead struct {
	Revision   int64            `json:"revision"`
	Status     enums.LeadStatus `json:"status"`
	OwnerID    int64            `json:"ownerId"`
	Result     string           `json:"result"`
	NextAction string           `json:"nextAction"`
	FollowUpAt *time.Time       `json:"followUpAt"`
}

File: web/lib/api/sales-lead.ts
import { request } from "@/lib/api/client";
import type { LeadStatus, LeadTag } from "@/lib/generated/enums";
import type { MemoryEntry } from "@/lib/api/conversation-memory";

export type LeadData = {
  title: string;
  product: string;
  quantity: string;
  destination: string;
  budget: string;
  timeline: string;
  contactName: string;
  company: string;
  email: string;
  phone: string;
  needs: string;
  autoTags: LeadTag[];
};
export type SalesLead = {
  id: number;
  conversationId: number;
  customerId: number;
  channelId: number;
  customerName: string;
  data: LeadData;
  proposal?: LeadData;
  confirmed: boolean;
  withdrawn: boolean;
  sourceValid: boolean;
  status: LeadStatus;
  ownerId: number;
  ownerName: string;
  customTags: string[];
  note: string;
  followUpAt?: string;
  nextAction: string;
  lastFollowUpAt?: string;
  revision: number;
  updatedAt: string;
  sources: MemoryEntry["sources"];
  events: { id: number; kind: string; actorId: number; createdAt: string; followUp?: LeadFollowUp }[];
};
export type LeadFollowUp = {
  revision: number;
  status: LeadStatus;
  ownerId: number;
  result: string;
  nextAction: string;
  followUpAt: string | null;
};
export function recordLeadFollowUp(id: number, payload: LeadFollowUp) {
  return request<void>(`/api/dashboard/conversation/sales_leads/${id}/follow_up`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}
export type LeadUpdate = {
  revision: number;
  data: LeadData;
  status: LeadStatus;
  ownerId: number;
  customTags: string[];
  note: string;
  followUpAt: string | null;
  acceptProposal: boolean;
};
export function fetchSalesLeads(
  query: Record<string, string | number | undefined> = {},
) {
  const params = new URLSearchParams();
  for (const [key, value] of Object.entries(query))
    if (value !== undefined && value !== "") params.set(key, String(value));
  return request<{
    results: SalesLead[];
    page: { page: number; limit: number; total: number };
  }>(`/api/dashboard/conversation/sales_leads?${params}`);
}
export function fetchSalesLead(id: number) {
  return request<SalesLead>(`/api/dashboard/conversation/sales_leads/${id}`);
}
export function updateSalesLead(id: number, payload: LeadUpdate) {
  return request<void>(`/api/dashboard/conversation/sales_leads/${id}/update`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}
